chain/chunks/src/lib.rs

```diff
@@ -178,53 +206,128 @@ impl SealsManager {
+            Entry::Vacant(entry) => {
+                let chunk_producer = self.runtime_adapter.get_chunk_producer(
+                    &self.runtime_adapter.get_epoch_id_from_prev_block(parent_hash)?,
+                    height,
+                    shard_id,
+                )?;
+                let candidates = {
+                    let n = self.runtime_adapter.num_total_parts();
+                    // `n` is an upper bound for elements in the accumulator; declaring with
+                    // this capacity up front will mean no further allocations will occur
+                    // from `push` calls in the loop.
+                    let mut accumulator = Vec::with_capacity(n);
+
+                    for part_ord in 0..n {
+                        let part_ord = part_ord as u64;
+                        let part_owner =
+                            self.runtime_adapter.get_part_owner(parent_hash, part_ord)?;
+                        if part_owner == chunk_producer || Some(part_owner) == self.me {
+                            continue;
+                        }
+                        accumulator.push(part_ord);
+                    }
+
+                    accumulator
+                };
+
+                let chosen = Self::get_random_part_ords(candidates);
+                let demur =
+                    ActiveSealDemur { part_ords: chosen, chunk_producer, sent: Utc::now(), height };
+
+                Ok(entry.insert(demur))
             }
-            let chosen = candidates
-                .choose_multiple(
-                    &mut rand::thread_rng(),
-                    cmp::min(NUM_PARTS_REQUESTED_IN_SEAL, candidates.len()),
-                )
-                .cloned()
-                .collect::<HashSet<_>>();
-            Seal { part_ords: chosen, chunk_producer, sent: Utc::now() }
-        }))
+        }
+    }
+
+    fn get_random_part_ords(candidates: Vec<u64>) -> HashSet<u64> {
+        candidates
+            .choose_multiple(
+                &mut rand::thread_rng(),
+                cmp::min(NUM_PARTS_REQUESTED_IN_SEAL, candidates.len()),
+            )
+            .cloned()
+            .collect()
     }
 
     fn approve_chunk(&mut self, chunk_hash: &ChunkHash) {
-        let seal = self.seals.get_mut(chunk_hash).expect("seal should be already produced");
-        seal.part_ords.clear();
+        let seal = self.active_demurs.remove(chunk_hash).expect("seal should be already produced");
+        Self::insert_past_seal(&mut self.past_seals, seal.height, chunk_hash.clone());
+    }
+
+    fn insert_past_seal(
+        past_seals: &mut BTreeMap<BlockHeight, HashSet<ChunkHash>>,
+        height: BlockHeight,
+        chunk_hash: ChunkHash,
+    ) {
+        let hashes_at_height = past_seals.entry(height).or_insert_with(HashSet::new);
+        hashes_at_height.insert(chunk_hash);
+    }
+
+    fn prune_past_seals(&mut self) {
+        let heights_to_remove = {
```

Comment:
why not use `split_off`? It is still slightly faster. It is O(m + logn) vs. O(m * logn) where m is the number of elements to remove.

Reply:
@bowenwang1996 @birchmd @SkidanovAlex I thought the whole discussion was about removing limited number of Seals, not limited number of Heights? I expect to see something like `while removed_seals < CONSTANT { self.past_seals[cur_height].remove(any_hash); ++removed_seals; }`

Reply:
@Kouprin  the version using the GC limit was also about removing some heights. If we wanted to keep it bounded by a fixed number of hashes then a `SizedCache` would work better. However, I don't think this is a good idea because the number of chunks produced at each height will depend on the number of shards, and I assume the number of shards will change over time. By using height to tag the hashes it allows us to measure how "old" they are using a metric that is more meaningful than actual time. And effectively the condition here is to remove hashes that are older than some fixed threshold.

Reply:
@bowenwang1996 true, good catch. I'll change it to use `split_off` as before.

Reply:
> @Kouprin the version using the GC limit was also about removing some heights. If we wanted to keep it bounded by a fixed number of hashes then a SizedCache would work better.

I'm not sure we're talking about the same. I didn't think about SizedCache. Let's say, there are heights 100, 101, 102, 103, 104; and `gc_height` is jumped from 100 to 105. There are some Chunks at Height 100 (a, b, c), some Chunks at Height 101 (d, e), some Chunks at Height 102 (f, g, h, ..., z, aa, ab, ..., zz). There are lots of Chunks at height 102, right? And only 3 Chunks at height 100 and 2 Chunks at height 101.

Now let's say we're limited to remove only 4 hashes per execution. We remove (a, b, c) from Height 101 and (d) from Height 102. Next time we will remove (e) from Height 102 and (f, g, h) from Height 103. Then next ~170 executions we will remove Chunks at Height 103 before moving to Height 104.

However, if we will use `split_off`, we will remove all the Chunks at Height 103 per one execution. There are lots of Chunks, ~26*27. It may be expensive. I expect we can solve it by limiting number of Chunk Hashes removed from `past_seals`, independently of how many Heights will be affected. (Sorry for using Seals previously in confusing manner).

Reply:
I don't think the cost of `split_off` changes significantly with the number of hashes at each height, so in your example, entirely removing 102 should be no more work than entirely removing 103.

The data structure is `BTreeMap<BlockHeight, HashSet<ChunkHash>>`, and `split_off` is a method on only the `BTreeMap`, so we never loop over any of the `HashSet`s. Ultimately we simply drop the side of the split with lower heights, which will cause the corresponding hash sets to be dropped as well, and this operation is `O(1)` I think because it is just marking memory has being freed.

Reply:
Got it. Can we mark this one as resolved?